### DISCUSSION TITLE
How do you set up Inventree in Cloud with Postgresql and https?

### DISCUSSION THREAD

**Discussion by dspmathguru** [inventree/InvenTree#6125] [Q&A] [answered]
I've been trying to setup having both postgresql and https (let's encrypt) and cannot find a good way to do this.

I have tried two methods, docker and bare metal and get stuck in both cases.

Docker is easy to get working with postgresql as it is the default.  I like this method, but cannot seem to figure out how to add certbot with let's encrypt.  It isn't obvious to me how to change the docker compose files to do this.  It seems to not want to create the cert.  If there is a pointer on how to do this, it would be great.  I am doing this on an AWS EC2 instance.  So in this case I can get things working without https, which I'm not willing to do with using this application.   A pointer to an example would be great that is known to work. 

Or ...

Next I tried the bare metal.  I do the by hand install as version 13 fails at the moment.  I get inventree working with sqlite3 and also I can get https working fine with let's encrypt.  (I use some of the pointers in this discussion thread to get it done).  The issue is when I try to change the config.yaml file with appropriate Posgresql changes, I can't seem to get the server to work.  So what are the best steps to bring up a different database once the standard sqlite3 setup is complete?

Basically I update /etc/inventree/config.yaml to use Postegresql.  Then what are you supposed to do?    If I do the following:
- cd /opt/inventree
- . env/bin/activate
- invoke update    <--- this fails as I think it updates to 13 which fails

I also tried to put the file in /etc/inventree/config.yaml and do a from scratch install of 12.8 with apt.  This fails as well.  I was thinking that maybe it used the config.yaml file as it was already there.  I'm not sure it is an issue here, but it is trying to back up data.  But since nothing has been created yet, it fails.  

What is the correct command for bare metal at this point?

It would be good to see a way to create a production server any pointers would be welcome.

**Comment by miggland** [comment, marked as answer]
One way to handle certificates and renewal is to use a reverse proxy, for example Traefik. It has plugins for most DNS providers, and is made to communicate with other Docker containers through labels.

First time I heard of it I thought "uhh, sounds complicated" - but once you get the hang of it, it makes sense. And it makes life easier if you have many containers on the same host.

**Reply by dspmathguru** [reply]
I will try it out.  Thank you.

**Reply by dspmathguru** [reply]
Thank you for the insight.  I used Traefik and it works great.   Here is my docker-compose.yml

```
version: "3.8"

# Docker compose recipe for a production-ready InvenTree setup, with the following containers:
# - PostgreSQL as the database backend
# - gunicorn as the InvenTree web server
# - django-q as the InvenTree background worker process
# - nginx as a reverse proxy
# - redis as the cache manager (optional, disabled by default)

# ---------------------
# READ BEFORE STARTING!
# ---------------------

# -----------------------------
# Setting environment variables
# -----------------------------
# Shared environment variables should be stored in the .env file
# Changes made to this file are reflected across all containers!
#
# IMPORTANT NOTE:
# You should not have to change *anything* within this docker-compose.yml file!
# Instead, make any changes in the .env file!

# ------------------------
# InvenTree Image Versions
# ------------------------
# By default, this docker-compose script targets the STABLE version of InvenTree,
# image: inventree/inventree:stable
#
# To run the LATEST (development) version of InvenTree,
# change the INVENTREE_TAG variable (in the .env file) to "latest"
#
# Alternatively, you could target a specific tagged release version with (for example):
# INVENTREE_TAG=0.7.5
#

services:
    traefik:
      image: "traefik:v2.10"
      container_name: "traefik"
      command:
          - "--log.level=DEBUG"
          - "--api.insecure=true"
          - "--providers.docker=true"
          - "--providers.docker.exposedbydefault=false"
          - "--entrypoints.web.address=:80"
          - "--entrypoints.web.http.redirections.entrypoint.to=websecure"
          - "--entrypoints.websecure.address=:443"
          - "--entrypoints.websecure.http.tls.domains[0].main=<WEBSITEURL>"
          - "--entrypoints.websecure.http.tls.domains[0].sans=*.<WEBSITEURL>"
          - "--certificatesresolvers.myresolver.acme.tlschallenge=true"
          - "--certificatesresolvers.myresolver.acme.httpchallenge.entrypoint=web"
          - "--certificatesresolvers.myresolver.acme.caserver=https://acme-v02.api.letsencrypt.org/directory"
          - "--certificatesresolvers.myresolver.acme.email=<EMAIL>"
          - "--certificatesresolvers.myresolver.acme.storage=/letsencrypt/acme.json"
      ports:
          - 443:443
          - 80:80
          #- 8000:8000
          #- ${INVENTREE_WEB_PORT:-1337}:${INVENTREE_WEB_PORT:-1337}
      volumes:
          - "inventree_data:/letsencrypt"
          - "/var/run/docker.sock:/var/run/docker.sock:ro"

    # Database service
    # Use PostgreSQL as the database backend
    inventree-db:
        image: postgres:13
        expose:
            - ${INVENTREE_DB_PORT:-5432}/tcp
        environment:
            - PGDATA=/var/lib/postgresql/data/pgdb
            - POSTGRES_USER=${INVENTREE_DB_USER:?You must provide the 'INVENTREE_DB_USER' variable in the .env file}
            - POSTGRES_PASSWORD=${INVENTREE_DB_PASSWORD:?You must provide the 'INVENTREE_DB_PASSWORD' variable in the .env file}
            - POSTGRES_DB=${INVENTREE_DB_NAME:?You must provide the 'INVENTREE_DB_NAME' variable in the .env file}
        volumes:
            # Map 'data' volume such that postgres database is stored externally
            - inventree_data:/var/lib/postgresql/data/
        restart: unless-stopped
        labels:
            - "traefik.enable=false"

    # redis acts as database cache manager
    # only runs under the "redis" profile : https://docs.docker.com/compose/profiles/
    inventree-cache:
        image: redis:7.0
        depends_on:
            - inventree-db
        profiles:
            - redis
        env_file:
            - .env
        expose:
            - ${INVENTREE_CACHE_PORT:-6379}
        restart: always
        labels:
            - "traefik.enable=false"

    # InvenTree web server service
    # Uses gunicorn as the web server
    inventree-server:
        # If you wish to specify a particular InvenTree version, do so here
        image: inventree/inventree:${INVENTREE_TAG:-stable}
        # Only change this port if you understand the stack.
        # If you change this you have to change:
        # - the proxy settings (on two lines)
        # - only change the exposed port - eg `1338:8000` if you want to expose the server on port 1338
        expose:
            - 8000
        depends_on:
            - inventree-db
        env_file:
            - .env
        volumes:
            # Data volume must map to /home/inventree/data
            - inventree_data:/home/inventree/data
        restart: unless-stopped
        labels:
            - "traefik.enable=false"

    # Background worker process handles long-running or periodic tasks
    inventree-worker:
        # If you wish to specify a particular InvenTree version, do so here
        image: inventree/inventree:${INVENTREE_TAG:-stable}
        command: invoke worker
        depends_on:
            - inventree-server
        env_file:
            - .env
        volumes:
            # Data volume must map to /home/inventree/data
            - inventree_data:/home/inventree/data
        restart: unless-stopped
        labels:
            - "traefik.enable=false"

    # nginx acts as a reverse proxy
    # static files are served directly by nginx
    # media files are served by nginx, although authentication is redirected to inventree-server
    # web requests are redirected to gunicorn
    # NOTE: You will need to provide a working nginx.conf file!
    inventree-proxy:
        image: nginx:stable
        depends_on:
            - inventree-server
        env_file:
            - .env
        labels:
            - "traefik.enable=true"
            - "traefik.http.routers.inventree-proxy.rule=Host(`<WEBSITE URL>`)"
            - "traefik.http.routers.inventree-proxy.entrypoints=websecure"
            - "traefik.http.routers.inventree-proxy.tls.certresolver=myresolver"
        ports:
            # Default web port is 1337 (can be changed in the .env file)
            - ${INVENTREE_WEB_PORT:-1337}:80
        volumes:
            # Provide nginx configuration file to the container
            # Refer to the provided example file as a starting point
            - ./nginx.prod.conf:/etc/nginx/conf.d/default.conf:ro
            # nginx proxy needs access to static and media files
            - inventree_data:/var/www
        restart: unless-stopped
volumes:
    # Persistent data, stored external to the container(s)
    inventree_data:
        driver: local
        driver_opts:
            type: none
            o: bind
            # This directory specified where InvenTree data are stored "outside" the docker containers
            device: ${INVENTREE_EXT_VOLUME:?You must specify the 'INVENTREE_EXT_VOLUME' variable in the .env file!}
```

**Reply by dspmathguru** [reply]
Wow, sorry for the crazy formatting.

**Reply by WanAqilDev** [reply]
Hi, good day.

Im quite new to this. if possible, may i ask some pointers to what direction should i go to understand the docker file better? i had just learned docker around a few weeks ago so i understand a bit. but the latest stable one uses caddy instead of nginx. how much differences does that make?

does adding traefik needs different configuration for caddy than nginx?

**Reply by dspmathguru** [reply]
I haven't updated to the latest docker-compose.yml, but I do not see why it wouldn't work with caddy.

**Reply by SchrodingersGat** [reply]
https://inventree.org/blog/2024/03/13/docker

**Reply by WanAqilDev** [reply]
Thank you so much for this guide. I've been riddled by the issues for some times now
